contrib/direnv

```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env bash
+
+# Look for the .config folder in the current directory and up
+find_prj_config() (
+	local old_pwd
+	while [[ $old_pwd != $PWD ]]; do
+		if [[ -d .config ]]; then
+			echo "$PWD"
+			return 0
+		fi
+		old_pwd=$PWD
+		cd ..
+	done
+	# We're at the top and didn't find anything
+	log_error "ERROR: could not find project root" >&2
+	return 1
+)
+
+{
+	: "${XDG_CONFIG_HOME:=${HOME}/.config}"
+	: "${XDG_CACHE_HOME:=${HOME}/.cache}"
+	: "${XDG_DATA_HOME:=${HOME}/.local/share}"
+
+	: "${GIT_DIR:=$(git rev-parse --absolute-git-dir 2>/dev/null)}"
+
+	# PRJ_ROOT
+	_prj_root="${GIT_DIR%/.git}"
```

Comment:
Hm, this points to a worktree, too:
```console
❯ git rev-parse --show-toplevel
/home/blaggacao/src/github.com/input-output-hk/ww
```

While the other command at least has the right repo in the path:
```console
❯ git rev-parse --absolute-git-dir
/home/blaggacao/src/github.com/input-output-hk/wallet-world/.git/worktrees/ww
```

Reply:
Fixed in 3ea5e756ca4c228aeb8586eb13fa899e295fdca6

Reply:
> Hm, this points to a worktree, too:
> 
> ```
> ❯ git rev-parse --show-toplevel
> /home/blaggacao/src/github.com/input-output-hk/ww
> ```
> 
> While the other command at least has the right repo in the path:
> 
> ```
> ❯ git rev-parse --absolute-git-dir
> /home/blaggacao/src/github.com/input-output-hk/wallet-world/.git/worktrees/ww
> ```

Sorry if I'm wrong here, but don't you want the former here? As I understand it, your goal is to get the root of the currently checked out git commit? 

I tried to describe the behaviour as I understand it below the break, but it's probably easier to show an example output instead. This should show that with the logic from 3ea5e756ca4c228aeb8586eb1 `PRJ_ROOT` always ends up pointing to the main checkout which would be wrong when working in a worktree.

```console
$ git init /tmp/wallet-world-0
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint:
hint:   git config --global init.defaultBranch <name>
hint:
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint:
hint:   git branch -m <name>
Initialized empty Git repository in /tmp/wallet-world-0/.git/
$ cd /tmp/wallet-world-0
$ touch some_file
$ git add some_file
$ git commit -m "Initial commit"
[master (root-commit) 8ef7fd9] Initial commit
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 some_file
$ git show --oneline
8ef7fd9 (HEAD -> master) Initial commit
diff --git a/some_file b/some_file
new file mode 100644
index 0000000..e69de29
$ export GIT_DIR=$(git rev-parse --absolute-git-dir)
$ export PRJ_ROOT="${GIT_DIR%/.git*}"
$ echo $GIT_DIR
/tmp/wallet-world-0/.git
$ echo $PRJ_ROOT
/tmp/wallet-world-0
$ ls $PRJ_ROOT
some_file
$ if [[ -e $PRJ_ROOT/some_new_file ]] ; then echo "on new commit"; else echo "on initial commit"; fi
on initial commit
$ unset GIT_DIR





$ git worktree add /tmp/wallet-world-1
Preparing worktree (new branch 'wallet-world-1')
HEAD is now at 8ef7fd9 Initial commit
$ cd /tmp/wallet-world-1
$ ls
some_file
$ touch some_new_file
$ git add some_new_file
$ git commit -m "New commit"
[wallet-world-1 851809f] New commit
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 some_new_file
$ git show --oneline
851809f (HEAD -> wallet-world-1) New commit
diff --git a/some_new_file b/some_new_file
new file mode 100644
index 0000000..e69de29
$ export GIT_DIR=$(git rev-parse --absolute-git-dir)
$ export PRJ_ROOT="${GIT_DIR%/.git*}"
$ echo $GIT_DIR
/tmp/wallet-world-0/.git/worktrees/wallet-world-1
$ echo $PRJ_ROOT
/tmp/wallet-world-0
$ ls $PRJ_ROOT
some_file
$ if [[ -e $PRJ_ROOT/some_new_file ]] ; then echo "on new commit"; else echo "on initial commit"; fi
on initial commit
$ unset GIT_DIR
```

I still think you need `--show-toplevel` which in this example would get you;

```console
$ cd /tmp/wallet-world-0
$ git rev-parse --show-toplevel
/tmp/wallet-world-0
$ cd /tmp/wallet-world-1
$ git rev-parse --show-toplevel
/tmp/wallet-world-1
```

**SOMETHING INTERESTING I DID NOTICE WHILE GENERATING ABOVE** is that setting `GIT_DIR` in the environment will override the behaviour of `git rev-parse --absolute-git-dir` and force it to use whatever is in that env var. I mention this since, while you don't export `GIT_DIR` in the script, it might cause some confusion if you've been experimenting with this stuff.

Hope that's helpful and that I haven't just added more confusion! 



<details>
<summary>More context</summary>


So if a standard checkout is currently on <some_commit> you'd have a file structure like;

```
/git/wallet-world-0/         
/git/wallet-world-0/.git      - git state directory
/git/wallet-world-0/some_file - file in the commit
```

assuming you want to work in this working directory you want `PRJ_ROOT` to be `/git/wallet-world-0` right?

Now say you want a new worktree on a different commit, and assuming you run `git worktree add /git/wallet-world-1 <some_new_commit>`you get this created;

```
/git/wallet-world-1/         
/git/wallet-world-1/.git           - git state. links to /git/wallet-world-0/.git/worktrees/wallet-world-1
/git/wallet-world-1/some_file      - file in the commit
/git/wallet-world-1/some_new_file  - some file exclusive to <some_commit>
```

If you want to work on this worktree on `<some_new_commit>` you presumably want `PRJ_ROOT` to be `/git/wallet-world-1`?

`git rev-parse --absolute-git-dir` should always give you the state directory, but for worktrees this state directory is not in the same place as the checkout. So it'd return something like `/git/wallet-world-0/.git/`.

But what I think you want is not the state directory but the actually checked out contents, i.e. `/git/wallet-world-1/`.


I guess as another example, say I have a script that checks which commit we're on;

```sh
if [[ -e $PRJ_ROOT/some_new_file ]] ; then
    echo "on some_new_commit"
else
    echo "on some_commit"
fi
```

when we run the script in `wallet-world-0` we should expect the output `on some_commit` and when we run it in `wallet-world-1` we should expect `on some_new_commit`. 

I think using the new logic from 3ea5e756ca4c228aeb8586eb1, if we run `git rev-parse --absolute-git-dir` in `wallet-world-1` we'll get the output `/git/wallet-world-0/.git/worktrees/wallet-world-1`.
If we then use the expansion `${GIT_DIR%/.git*}` we'll end up with `/git/wallet-world-0/` so even though we're running in `wallet-world-1` `PRJ_ROOT` is going to point at `wallet-world-0`. 


</details>

Reply:
One question:

Do we want:
- `PRJ_ROOT` point to worktree checkout; yet
- `PRJ_CONFIG_HOME` & al point to origin checkout (where `./.config` & al already live)

?

cc @zimbatm

Reply:
Given that, with `PRJ_ID`, we have an escape hatch to lift the state into its XDG equivalents, it captures the intent of global shared state for reason XYZ.

With worktrees, users aren't hindered to do so as well.

Hence, you are right. We shall consider them separately.

Reply:
440705abeb8de9e63c703724a21802c89f6600c7